File: ios/kls_gps_tracker/Sources/kls_gps_tracker/KlsGpsStorage.swift
import CoreLocation
import Foundation

final class KlsGpsStorage {
  private let fileManager = FileManager.default
  private let defaults = UserDefaults.standard
  private let lock = NSLock()

  private lazy var storageDirectory: URL = {
    let base = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
    var directory = base.appendingPathComponent("kls_gps_tracker", isDirectory: true)
    try? fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
    var resourceValues = URLResourceValues()
    resourceValues.isExcludedFromBackup = true
    try? directory.setResourceValues(resourceValues)
    return directory
  }()

  func beginWorkout(requestedWorkoutId: String?) throws -> String {
    lock.lock()
    defer { lock.unlock() }

    let requested = requestedWorkoutId?.trimmingCharacters(in: .whitespacesAndNewlines)
    if let requested, !requested.isEmpty {
      try validateWorkoutId(requested)
    }

    let active = activeWorkoutId
    if let active, let requested, !requested.isEmpty, active != requested {
      throw storageError("Workout \(active) is already being recorded.")
    }

    let workoutId = active ?? ((requested?.isEmpty == false) ? requested! : UUID().uuidString.lowercased())
    defaults.set(workoutId, forKey: Keys.activeWorkoutId)
    defaults.set(true, forKey: Keys.shouldAutoResume)
    _ = ensureNextPointIndex(workoutId)
    return workoutId
  }

  var activeWorkoutId: String? {
    defaults.string(forKey: Keys.activeWorkoutId)?.nilIfEmpty
  }

  var isTracking: Bool {
    defaults.bool(forKey: Keys.isTracking)
  }

  func setTracking(_ value: Bool) {
    defaults.set(value, forKey: Keys.isTracking)
  }

  var shouldAutoResume: Bool {
    defaults.bool(forKey: Keys.shouldAutoResume)
  }

  func pauseWorkout() {
    defaults.set(false, forKey: Keys.isTracking)
    defaults.set(false, forKey: Keys.shouldAutoResume)
  }

  func finishWorkout(_ workoutId: String?) {
    lock.lock()
    defer { lock.unlock() }

    defaults.set(false, forKey: Keys.isTracking)
    defaults.set(false, forKey: Keys.shouldAutoResume)
    if workoutId == nil || activeWorkoutId == workoutId {
      defaults.removeObject(forKey: Keys.activeWorkoutId)
    }
  }

  func append(location: CLLocation, workoutId: String) throws -> [String: Any] {
    lock.lock()
    defer { lock.unlock() }

    try validateWorkoutId(workoutId)
    let pointIndex = ensureNextPointIndex(workoutId)
    var point: [String: Any] = [
      "workoutId": workoutId,
      "pointIndex": pointIndex,
      "latitude": location.coordinate.latitude,
      "longitude": location.coordinate.longitude,
      "accuracy": location.horizontalAccuracy,
      "timestampMillis": location.timestamp.timeIntervalSince1970 * 1000,
      "provider": "coreLocation",
      "isMock": false,
    ]

    if location.verticalAccuracy >= 0 {
      point["altitude"] = location.altitude
      point["verticalAccuracy"] = location.verticalAccuracy
    }
    if location.speed >= 0 {
      point["speed"] = location.speed
      point["speedAccuracy"] = location.speedAccuracy
    }
    if location.course >= 0 {
      point["heading"] = location.course
      if #available(iOS 13.4, *) {
        point["headingAccuracy"] = location.courseAccuracy
      }
    }
    if #available(iOS 15.0, *) {
      point["isMock"] = location.sourceInformation?.isSimulatedBySoftware == true
    }

    var encoded = try JSONSerialization.data(withJSONObject: point)
    encoded.append(0x0A)
    let fileURL = fileURL(for: workoutId)
    if !fileManager.fileExists(atPath: fileURL.path) {
      fileManager.createFile(atPath: fileURL.path, contents: nil)
    }
    let handle = try FileHandle(forWritingTo: fileURL)
    handle.seekToEndOfFile()
    handle.write(encoded)
    handle.synchronizeFile()
    handle.closeFile()

    defaults.set(pointIndex + 1, forKey: nextIndexKey(workoutId))
    return point
  }

  func readPoints(
    workoutId: String,
    afterPointIndex: Int,
    limit: Int
  ) throws -> [[String: Any]] {
    try validateWorkoutId(workoutId)
    let fileURL = fileURL(for: workoutId)
    guard fileManager.fileExists(atPath: fileURL.path) else { return [] }

    let raw = try String(contentsOf: fileURL, encoding: .utf8)
    var result: [[String: Any]] = []
    let safeLimit = min(max(limit, 1), 5000)
    for line in raw.split(separator: "\n", omittingEmptySubsequences: true) {
      guard
        let data = String(line).data(using: .utf8),
        let point = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
        let pointIndex = (point["pointIndex"] as? NSNumber)?.intValue,
        pointIndex > afterPointIndex
      else {
        continue
      }
      result.append(point)
      if result.count >= safeLimit { break }
    }
    return result
  }

  func pointCount(workoutId: String) -> Int {
    lock.lock()
    defer { lock.unlock() }
    return ensureNextPointIndex(workoutId)
  }

  func listStoredWorkoutIds() -> [String] {
    let files = (try? fileManager.contentsOfDirectory(
      at: storageDirectory,
      includingPropertiesForKeys: nil
    )) ?? []
    return files
      .filter { $0.pathExtension == "jsonl" }
      .map { $0.deletingPathExtension().lastPathComponent }
      .sorted()
  }

  func deleteWorkout(_ workoutId: String) throws {
    lock.lock()
    defer { lock.unlock() }

    try validateWorkoutId(workoutId)
    if activeWorkoutId == workoutId {
      throw storageError("The active workout cannot be deleted.")
    }
    let fileURL = fileURL(for: workoutId)
    if fileManager.fileExists(atPath: fileURL.path) {
      try fileManager.removeItem(at: fileURL)
    }
    defaults.removeObject(forKey: nextIndexKey(workoutId))
  }

  private func ensureNextPointIndex(_ workoutId: String) -> Int {
    let key = nextIndexKey(workoutId)
    if defaults.object(forKey: key) != nil {
      return defaults.integer(forKey: key)
    }

    let fileURL = fileURL(for: workoutId)
    let count: Int
    if let raw = try? String(contentsOf: fileURL, encoding: .utf8) {
      count = raw.split(separator: "\n", omittingEmptySubsequences: true).count
    } else {
      count = 0
    }
    defaults.set(count, forKey: key)
    return count
  }

  private func fileURL(for workoutId: String) -> URL {
    storageDirectory.appendingPathComponent("\(workoutId).jsonl", isDirectory: false)
  }

  private func nextIndexKey(_ workoutId: String) -> String {
    "kls_gps_next_index_\(workoutId)"
  }

  private func validateWorkoutId(_ workoutId: String) throws {
    let range = NSRange(location: 0, length: workoutId.utf16.count)
    guard Self.workoutIdRegex.firstMatch(in: workoutId, range: range) != nil else {
      throw storageError(
        "workoutId must contain only letters, digits, dots, underscores, or hyphens."
      )
    }
  }

  private func storageError(_ message: String) -> NSError {
    NSError(
      domain: "com.kls.kls_gps_tracker.storage",
      code: 1,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }

  private static let workoutIdRegex = try! NSRegularExpression(
    pattern: "^[A-Za-z0-9._-]{1,128}$"
  )

  private enum Keys {
    static let activeWorkoutId = "kls_gps_active_workout_id"
    static let isTracking = "kls_gps_is_tracking"
    static let shouldAutoResume = "kls_gps_should_auto_resume"
  }
}

private extension String {
  var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: android/src/main/kotlin/com/kls/kls_gps_tracker/KlsGpsStorage.kt
package com.kls.kls_gps_tracker

import android.content.Context
import android.location.Location
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.nio.charset.StandardCharsets
import java.util.UUID

internal class KlsGpsStorage(context: Context) {
    private val appContext = context.applicationContext
    private val preferences =
        appContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
    private val storageDirectory = File(appContext.filesDir, STORAGE_DIRECTORY).apply { mkdirs() }

    @Synchronized
    fun beginWorkout(requestedWorkoutId: String?): String {
        val requested = requestedWorkoutId?.trim()?.takeIf { it.isNotEmpty() }
        if (requested != null) validateWorkoutId(requested)

        val active = activeWorkoutId()
        if (active != null && requested != null && active != requested) {
            throw IllegalStateException("Workout $active is already being recorded.")
        }

        val workoutId = active ?: requested ?: UUID.randomUUID().toString()
        preferences
            .edit()
            .putString(KEY_ACTIVE_WORKOUT_ID, workoutId)
            .putBoolean(KEY_SHOULD_AUTO_RESUME, true)
            .apply()
        ensureNextPointIndex(workoutId)
        return workoutId
    }

    fun activeWorkoutId(): String? =
        preferences.getString(KEY_ACTIVE_WORKOUT_ID, null)?.takeIf { it.isNotBlank() }

    fun isTracking(): Boolean = preferences.getBoolean(KEY_IS_TRACKING, false)

    fun setTracking(value: Boolean) {
        preferences.edit().putBoolean(KEY_IS_TRACKING, value).apply()
    }

    fun shouldAutoResume(): Boolean =
        preferences.getBoolean(KEY_SHOULD_AUTO_RESUME, false)

    fun pauseWorkout() {
        preferences
            .edit()
            .putBoolean(KEY_IS_TRACKING, false)
            .putBoolean(KEY_SHOULD_AUTO_RESUME, false)
            .apply()
    }

    @Synchronized
    fun finishWorkout(workoutId: String?) {
        val active = activeWorkoutId()
        val editor =
            preferences
                .edit()
                .putBoolean(KEY_IS_TRACKING, false)
                .putBoolean(KEY_SHOULD_AUTO_RESUME, false)
        if (workoutId == null || active == workoutId) {
            editor.remove(KEY_ACTIVE_WORKOUT_ID)
        }
        editor.apply()
    }

    @Synchronized
    fun appendLocation(workoutId: String, location: Location): Map<String, Any> {
        validateWorkoutId(workoutId)
        val pointIndex = ensureNextPointIndex(workoutId)
        val point = linkedMapOf<String, Any>(
            "workoutId" to workoutId,
            "pointIndex" to pointIndex,
            "latitude" to location.latitude,
            "longitude" to location.longitude,
            "accuracy" to if (location.hasAccuracy()) location.accuracy.toDouble() else -1.0,
            "timestampMillis" to location.time,
            "provider" to (location.provider ?: "unknown"),
            "isMock" to location.isFromMockProvider,
            "elapsedRealtimeNanos" to location.elapsedRealtimeNanos,
        )

        if (location.hasAltitude()) point["altitude"] = location.altitude
        if (location.hasSpeed()) point["speed"] = location.speed.toDouble()
        if (location.hasBearing()) point["heading"] = location.bearing.toDouble()
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            if (location.hasVerticalAccuracy()) {
                point["verticalAccuracy"] = location.verticalAccuracyMeters.toDouble()
            }
            if (location.hasSpeedAccuracy()) {
                point["speedAccuracy"] = location.speedAccuracyMetersPerSecond.toDouble()
            }
            if (location.hasBearingAccuracy()) {
                point["headingAccuracy"] = location.bearingAccuracyDegrees.toDouble()
            }
        }

        val encoded = JSONObject(point as Map<*, *>).toString() + "\n"
        FileOutputStream(fileFor(workoutId), true).use { stream ->
            stream.write(encoded.toByteArray(StandardCharsets.UTF_8))
            stream.flush()
            stream.fd.sync()
        }

        preferences.edit().putInt(nextIndexKey(workoutId), pointIndex + 1).apply()
        return point
    }

    @Synchronized
    fun readPoints(
        workoutId: String,
        afterPointIndex: Int,
        limit: Int,
    ): List<Map<String, Any>> {
        validateWorkoutId(workoutId)
        val file = fileFor(workoutId)
        if (!file.exists()) return emptyList()

        val result = ArrayList<Map<String, Any>>(limit.coerceAtMost(5000))
        file.bufferedReader(StandardCharsets.UTF_8).useLines { lines ->
            for (line in lines) {
                if (line.isBlank()) continue
                val json = runCatching { JSONObject(line) }.getOrNull() ?: continue
                val pointIndex = json.optInt("pointIndex", -1)
                if (pointIndex <= afterPointIndex) continue

                val point = linkedMapOf<String, Any>()
                val keys = json.keys()
                while (keys.hasNext()) {
                    val key = keys.next()
                    val value = json.opt(key)
                    if (value != null && value !== JSONObject.NULL) point[key] = value
                }
                result.add(point)
                if (result.size >= limit.coerceIn(1, 5000)) break
            }
        }
        return result
    }

    fun pointCount(workoutId: String): Int {
        validateWorkoutId(workoutId)
        return ensureNextPointIndex(workoutId)
    }

    fun listStoredWorkoutIds(): List<String> =
        storageDirectory
            .listFiles { file -> file.isFile && file.extension == FILE_EXTENSION }
            ?.map { it.nameWithoutExtension }
            ?.sorted()
            ?: emptyList()

    @Synchronized
    fun deleteWorkout(workoutId: String) {
        validateWorkoutId(workoutId)
        if (activeWorkoutId() == workoutId) {
            throw IllegalStateException("The active workout cannot be deleted.")
        }
        fileFor(workoutId).delete()
        preferences.edit().remove(nextIndexKey(workoutId)).apply()
    }

    private fun ensureNextPointIndex(workoutId: String): Int {
        val key = nextIndexKey(workoutId)
        if (preferences.contains(key)) return preferences.getInt(key, 0)

        val file = fileFor(workoutId)
        val count = if (file.exists()) {
            file.bufferedReader(StandardCharsets.UTF_8).useLines { lines ->
                lines.count { it.isNotBlank() }
            }
        } else {
            0
        }
        preferences.edit().putInt(key, count).apply()
        return count
    }

    private fun fileFor(workoutId: String): File =
        File(storageDirectory, "$workoutId.$FILE_EXTENSION")

    private fun nextIndexKey(workoutId: String): String = "next_index_$workoutId"

    private fun validateWorkoutId(workoutId: String) {
        require(WORKOUT_ID_PATTERN.matches(workoutId)) {
            "workoutId must contain only letters, digits, dots, underscores, or hyphens."
        }
    }

    private companion object {
        const val PREFERENCES_NAME = "kls_gps_tracker"
        const val STORAGE_DIRECTORY = "kls_gps_tracker"
        const val FILE_EXTENSION = "jsonl"
        const val KEY_ACTIVE_WORKOUT_ID = "active_workout_id"
        const val KEY_IS_TRACKING = "is_tracking"
        const val KEY_SHOULD_AUTO_RESUME = "should_auto_resume"
        val WORKOUT_ID_PATTERN = Regex("[A-Za-z0-9._-]{1,128}")
    }
}
